File: activityloghelper/src/main/java/com/example/administrator/activityloghelper/model/AccessibilityViewModel.java
package com.example.administrator.activityloghelper.model;

import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleObserver;
import android.arch.lifecycle.LifecycleOwner;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.OnLifecycleEvent;
import android.arch.lifecycle.ViewModel;
import android.content.Context;
import android.util.Log;
import android.view.accessibility.AccessibilityManager;

import com.example.administrator.activityloghelper.R;


public class AccessibilityViewModel extends ViewModel implements LifecycleObserver, AccessibilityManager.AccessibilityStateChangeListener {

    private Context applicationContext;

    private MutableLiveData<String> mutableLiveData = new MutableLiveData<>();


    public AccessibilityViewModel(Context context) {
        applicationContext = context.getApplicationContext();
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    public void onCreate() {
        //监听AccessibilityService 变化
        AccessibilityManager accessibilityManager = (AccessibilityManager) applicationContext.getSystemService(Context.ACCESSIBILITY_SERVICE);
        if (accessibilityManager != null) {
            accessibilityManager.addAccessibilityStateChangeListener(this);
        }
    }


    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    public void onDestroy() {
        //移除监听服务
        AccessibilityManager accessibilityManager = (AccessibilityManager) applicationContext.getSystemService(Context.ACCESSIBILITY_SERVICE);
        if (accessibilityManager != null) {
            accessibilityManager.removeAccessibilityStateChangeListener(this);
        }
    }

    private static final String TAG = "wangyi";

    @Override
    public void onAccessibilityStateChanged(boolean enabled) {
        Log.d(TAG, "onAccessibilityStateChanged: ");
        mutableLiveData.setValue(applicationContext.getString(enabled ? R.string.service_off : R.string.service_on));
    }

    public LiveData<String> getChangeLiveData() {
        return mutableLiveData;
    }
}

File: activityloghelper/src/main/java/com/example/administrator/activityloghelper/model/AccessibilityViewModel.kt
package com.example.administrator.activityloghelper.model;

import android.content.Context
import android.view.accessibility.AccessibilityManager
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.airbnb.mvrx.MvRxState
import com.airbnb.mvrx.MvRxViewModelFactory
import com.airbnb.mvrx.ViewModelContext
import com.example.administrator.activityloghelper.LogHelperApplication
import com.example.administrator.activityloghelper.MvRxViewModel
import com.example.administrator.activityloghelper.R

data class AccessibilityState(val status: String = "", val windowStatus: String = "") :
    MvRxState

class AccessibilityViewModel(
    accessibilityState: AccessibilityState,
    private val appContext: Context
) : MvRxViewModel<AccessibilityState>(accessibilityState),
    AccessibilityManager.AccessibilityStateChangeListener {


    init {
        //监听AccessibilityService 变化
        val accessibilityManager =
            appContext.getSystemService(Context.ACCESSIBILITY_SERVICE) as AccessibilityManager
        accessibilityManager.addAccessibilityStateChangeListener(this)
    }


    override fun onCleared() {
        super.onCleared()
        //监听AccessibilityService 变化
        val accessibilityManager =
            appContext.getSystemService(Context.ACCESSIBILITY_SERVICE) as AccessibilityManager
        accessibilityManager.removeAccessibilityStateChangeListener(this)
    }

    override fun onAccessibilityStateChanged(enabled: Boolean) {
        setState {
            val result: String = if (enabled) {
                appContext.getString(R.string.service_off)
            } else {
                appContext.getString(R.string.service_on)
            }
            copy(status = result)
        }
    }

    fun updateWindowStatus(isShowing: Boolean) {
        setState {
            val text: String = if (isShowing) {
                appContext.getString(R.string.window_off)
            } else {
                appContext.getString(R.string.window_on)
            }
            copy(windowStatus = text)
        }
    }

    companion object : MvRxViewModelFactory<AccessibilityViewModel, AccessibilityState> {

        override fun create(
            viewModelContext: ViewModelContext,
            state: AccessibilityState
        ): AccessibilityViewModel? {
            val app = viewModelContext.app<LogHelperApplication>()
            return AccessibilityViewModel(state, app)
        }

    }
}
